src/Analyser/MutatingScope.php

```diff
@@ -3803,11 +3806,32 @@ public function filterBySpecifiedTypes(SpecifiedTypes $specifiedTypes): self
 			return $b['sure'] - $a['sure']; // @phpstan-ignore-line
 		});
 
-		$scope = $this;
 		$specifiedExpressions = [];
 		foreach ($typeSpecifications as $typeSpecification) {
 			$expr = $typeSpecification['expr'];
 			$type = $typeSpecification['type'];
+
+			if ($expr instanceof NotIssetExpr) {
+				$unsetExpr = $expr->getExpr();
```

Comment:
Do this only for 'sure'=true.

Reply:
moving it into the `'sure'=true.` branch regresses some tests. I think what you want to achieve is already taken care of in the `TypeSpecifier` side of this PR, as `NotIssetExpr` is only passed thru in the [`!$context->true()`](https://github.com/phpstan/phpstan-src/pull/2657/files#diff-1076e38598f300a69c9dc998dae31374052f95bd69d0d71c11e9d0a4c188976dR678)-case